lib/generators/deimos/active_record_consumer_generator.rb

```diff
@@ -0,0 +1,151 @@
+          template('consumer.rb', "app/lib/kafka/models/#{consumer_name.underscore}.rb")
+        end
+
+        # Adds consumer config to config file.
+        # Defaults to deimos.rb if config_path arg is not specified.
+        def create_consumer_config
+          config_file = 'deimos.rb'
+          config_file_path = "#{initializer_path}/#{config_file}"
+          config_file_path = config_path if config_path.present?
+
+          if File.exist?(config_file_path)
+            config_template = File.expand_path(find_in_source_paths('config.rb'))
+            insert_into_file(config_file_path.to_s, CapturableERB.new(::File.binread(config_template)).result(binding))
+          else
+            template('config.rb', config_file_path.to_s)
+          end
+        end
+
+        # Generates schema classes
+        def create_deimos_schema_class
+          Deimos::Generators::SchemaClassGenerator.start
+        end
+
+        # @return [String]
+        def db_migrate_path
+          if defined?(Rails.application) && Rails.application
+            paths = Rails.application.config.paths['db/migrate']
+            paths.respond_to?(:to_ary) ? paths.to_ary.first : paths.to_a.first
+          else
+            'db/migrate'
+          end
+        end
+
+        # @return [String]
+        def migration_version
+          "[#{ActiveRecord::Migration.current_version}]"
+        rescue StandardError
+          ''
+        end
+
+        # @return [String]
+        def table_class
+          table_name.classify
+        end
+
+        # @return [String]
+        def schema
+          last_dot = self.full_schema.rindex('.')
+          self.full_schema[last_dot + 1..-1]
+        end
+
+        # @return [String]
+        def namespace
+          last_dot = self.full_schema.rindex('.')
+          self.full_schema[0...last_dot]
+        end
+
+        # @return [Deimos::SchemaBackends::Base]
+        def schema_base
+          @schema_base ||= Deimos.schema_backend_class.new(schema: schema, namespace: namespace)
+        end
+
+        # @return [Array<SchemaField>]
+        def fields
+          schema_base.schema_fields
+        end
+
+        # @return [String]
+        def table_name
+          schema.tableize
+        end
+
+        # @return [String]
+        def model_name
+          table_name.underscore.singularize
+        end
+
+        # @return [String]
+        def consumer_name
+          "#{schema.classify}Consumer"
+        end
+
+        # @return [String]
+        def initializer_path
+          if defined?(Rails.application) && Rails.application
+            paths = Rails.application.config.paths['config/initializers']
+            paths.respond_to?(:to_ary) ? paths.to_ary.first : paths.to_a.first
+          else
+            'config/initializers'
+          end
+        end
+
+        # @return [String] Returns the name of the first field in the schema, as the key
+        def key_field
+          fields.first.name
```

Comment:
This is almost definitely not going to be the right approach. It probably makes more sense to just have the key config specified as part of the input. We'll also need the key schema (if there is one) when generating the schema classes.

Reply:
Ok, yes makes sense to include key config as an input.

In terms of the key schema, how does that work? I noticed that the schema glasses generator doesn't take in any arguments. So, if the user has a key schema for the new consumer, do they just have to place it in the SCHEMA_ROOT prior to running the ActiveRecordConsumer generator? And then what will be the difference in output? Will the generated schema classes be different?

Reply:
The generator reads the configuration which already knows what the key schema is.

The generator has a method `generate_classes(schema_name, namespace, key_config)` which you should probably be using.